sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -166,46 +183,122 @@ class GpuBatchSubPartitioner(
-        // 2) Split into smaller tables according to partitions
-        val subTables = withResource(partedTable) { _ =>
-          partedTable.getTable.contiguousSplit(partedTable.getPartitions.tail: _*)
-        }
-        // 3) Make each smaller table spillable and cache them in the queue
-        withResource(subTables) { _ =>
-          subTables.zipWithIndex.foreach { case (table, id) =>
-            // skip empty tables
-            if (table.getRowCount > 0) {
-              pendingParts(id) += SpillableColumnarBatch(table, types,
-                SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-              numCurBatches += 1
+  protected def partitionBatches(numParts: Int): SubPartitionBuffer = {
+    partitionBatchesOnceAndClose(inputIter, numParts)
+  }
+
+  protected final def partitionBatchesOnceAndClose(batchesIter: Iterator[ColumnarBatch],
+      numParts: Int): SubPartitionBuffer = {
+    closeOnExcept(Array.fill(numParts)(ArrayBuffer.empty[SpillableColumnarBatch])) { subParts =>
+      val hashSeed = seedGenerator.nextSeed
+      while (batchesIter.hasNext) {
+        val gpuBatch = batchesIter.next()
+        if (gpuBatch.numRows() > 0 && gpuBatch.numCols() > 0) {
+          val types = closeOnExcept(gpuBatch)(GpuColumnVector.extractTypes)
+          // 1) Hash partition on the batch
+          val partedTable = GpuHashPartitioningBase.hashPartitionAndClose(gpuBatch,
+            inputBoundKeys, numParts, "Sub-Hash Calculate", hashSeed)
+          // 2) Split into smaller tables according to partitions
+          val subTables = withResource(partedTable) { _ =>
+            partedTable.getTable.contiguousSplit(partedTable.getPartitions.tail: _*)
+          }
+          // 3) Make each smaller table spillable and cache them in the queue
+          withResource(subTables) { _ =>
+            subTables.zipWithIndex.foreach { case (table, id) =>
+              // skip empty tables
+              if (table.getRowCount > 0) {
+                subParts(id) += SpillableColumnarBatch(table, types,
+                  SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+              }
             }
           }
+        } else if (gpuBatch.numRows() > 0 && gpuBatch.numCols() == 0) {
+          // Rows only batch. This should never happen for a hash join in Spark.
+          gpuBatch.close()
+        } else {
+          // Skip empty batches
+          gpuBatch.close()
         }
-      } else if (gpuBatch.numRows() > 0 && gpuBatch.numCols() == 0) {
-        // Rows only batch. This should never happen for a hash join in Spark.
-        gpuBatch.close()
-      } else {
-        // Skip empty batches
-        gpuBatch.close()
       }
-    } // end of while
+      subParts
+    }
+  }
+
+  override final def currentSeed: Int = seedGenerator.currentSeed
+
+  /**
+   * Always return the current seed.
+   * This is intended to share the same seed across sub-partitioners.
+   */
+  override final def nextSeed: Int = seedGenerator.currentSeed
```

Comment:
nit: This is confusing. I would prefer it if we had a way to pass the seed directly into a sub-partitioner instead.

Reply:
Done